### PR TITLE
Added a basic game list

### PR DESCRIPTION
**Pull request by SteveLeafo**
A port of the Yuzu game list to decaf.

### CODE REVIEW

**Comment by FearlessTobi** [comment]
Screenshot would be cool.

File: src/decaf-qt/gamelist.cpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+   mItemModel->setHeaderData(COLUMN_NAME, Qt::Horizontal, tr("Name"));
+   mItemModel->setHeaderData(COLUMN_COMPATIBILITY, Qt::Horizontal, tr("Compatibility"));  
+   mItemModel->setHeaderData(COLUMN_FILE_TYPE - 1, Qt::Horizontal, tr("Title ID"));
+   mItemModel->setHeaderData(COLUMN_SIZE - 1, Qt::Horizontal, tr("Path"));
+
+   connect(mTreeView, &QTreeView::activated, this, &GameList::validateEntry);
+   connect(mWatcher, &QFileSystemWatcher::directoryChanged, this, &GameList::refreshGameDirectory);
+
+   // We must register all custom types with the Qt Automoc system so that we are able to use it
+   // with signals/slots. In this case, QList falls under the umbrells of custom types.
```
Comment (VelocityRa) [inline src/decaf-qt/gamelist.cpp]:
```suggestion
   // with signals/slots. In this case, QList falls under the umbrellas of custom types.
```

File: src/decaf-qt/gamelist.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+      const QDir dir{filePath};
+      const QStringList matchingMain = dir.entryList(QStringList("main"), QDir::Files);
+
+      if (matchingMain.size() == 1) {
+         emit gameChosen(dir.path() + "/" + matchingMain[0]);
+      }
+      return;
+   }
+
+   // Users usually want to run a diffrent game after closing one
```
Comment (VelocityRa) [inline src/decaf-qt/gamelist.cpp]:
```suggestion
   // Users usually want to run a different game after closing one
```

File: src/decaf-qt/gamelistp.h [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      static const std::map<QString, CompatStatus> statusData = 
+      {
+         {"0",  {"#5c93ed", QT_TR_NOOP("Perfect"),   QT_TR_NOOP("Game functions flawless with no audio or graphical glitches, all tested functionality works as intended without\nany workarounds needed.")}},
+         {"1",  {"#47d35c", QT_TR_NOOP("Great"),     QT_TR_NOOP("Game functions with minor graphical or audio glitches and is playable from start to finish. May require some\nworkarounds.")}},
+         {"2",  {"#94b242", QT_TR_NOOP("Okay"),      QT_TR_NOOP("Game functions with major graphical or audio glitches, but game is playable from start to finish with\nworkarounds.")}},
+         {"3",  {"#f2d624", QT_TR_NOOP("Bad"),      QT_TR_NOOP("Game functions, but with major graphical or audio glitches. Unable to progress in specific areas due to glitches\neven with workarounds.")}},
+         {"4",  {"#FF0000", QT_TR_NOOP("Intro/Menu"), QT_TR_NOOP("Game is completely unplayable due to major graphical or audio glitches. Unable to progress past the Start\nScreen.")}},
+         {"5",  {"#828282", QT_TR_NOOP("Won't Boot"), QT_TR_NOOP("The game crashes when attempting to startup.")}},
```
Comment (VelocityRa) [inline src/decaf-qt/gamelistp.h]:
```suggestion
         {"5",  {"#828282", QT_TR_NOOP("Won't Boot"), QT_TR_NOOP("The game crashes when attempting to start up.")}},
```

File: src/decaf-qt/settingsdialog/displaysettingswidget.cpp [hunk 4]
```diff
@@ -38,4 +40,14 @@ DisplaySettingsWidget::saveSettings(Settings &settings)
...
 }
+
+void
+DisplaySettingsWidget::browseGamePath()
+{
+   auto path = QFileDialog::getExistingDirectory(this, tr("Open Directory"), mUi.lineEditGamePaths->text());
+   if (!path.isEmpty()) {
+      mUi.lineEditGamePaths->setText(path);
+   }
+}
```
Comment (VelocityRa) [inline src/decaf-qt/settingsdialog/displaysettingswidget.cpp]: newline

**Comment by exjam** [comment]
1. I think we should remove the "Compatibility" rating from the game list - a compatibility list is too hard to maintain accurately, and should not be part of the emulator itself imo.

2. Regarding the icons, you should be able to use `meta/iconTex.tga`.

3. The setting gamePath should probably be renamed to titlePath.

4. The setting gamePath/titlePath should probably be part of the system settings in [decaf_config.h](https://github.com/decaf-emu/decaf-emu/blob/master/src/libdecaf/decaf_config.h) and be a `std::vector<std::string>`.

5. You should verify that titles found in titlePaths are actually games and not updates or DLC (so that we can use title path to find DLC + Updates for a loaded game). I believe you can do this just by checking the lowest 4 bits of the highest 32bits of title id as per the following:
```
(title_id_hi & 0xF) == 0x0 = Title 👍 
(title_id_hi & 0xF) == 0x2 = Demo 👍 
(title_id_hi & 0xF) == 0xB = Data 👎 
(title_id_hi & 0xF) == 0xC = DLC 👎 
(title_id_hi & 0xF) == 0xE = Update 👎 
```
ref: http://wiiubrew.org/wiki/Title_database

File: src/decaf-qt/gamelistworker.cpp [hunk 5]
```diff
@@ -222,6 +172,7 @@ ForeachDirectoryEntry(uint64_t* numEntriesOut,
    // windows loop
```
Comment (exjam) [inline src/decaf-qt/gamelistworker.cpp]: Should be using Qt to do this.

http://doc.qt.io/qt-5/qdiriterator.html
http://doc.qt.io/qt-5/examples-xml.html
http://doc.qt.io/qt-5/qxmlsimplereader.html

**Comment by exjam** [comment]
2c2e493007d9cc9737fe47820a4dc3aa63c10ded

**Pull request closed without merging**